_posts/2022-08-22-fan-out-and-back-in-using-functions.md

```diff
@@ -0,0 +1,415 @@
+
+The code of the `collect-result` function:
+
+```python
+import os
+import json
+import boto3
+
+from smart_open import s3
+from smart_open import open
+
+def handle(event, context):
+    bucketName = os.getenv('s3_bucket')
+
+    with open('/var/openfaas/secrets/s3-key', 'r') as s:
+        s3Key = s.read()
+    with open('/var/openfaas/secrets/s3-secret', 'r') as s:
+        s3Secret = s.read()
+
+    session = boto3.Session(
+        aws_access_key_id=s3Key,
+        aws_secret_access_key=s3Secret,
+    )
+
+    batchId = event.headers.get('X-Batch-Id')
+
+    passed = []
+    failed = []
+    
+    for key, content in s3.iter_bucket(bucketName, prefix=batchId + '/', workers=30, aws_access_key_id=s3Key, aws_secret_access_key=s3Secret):
+        data = json.loads(content)
+        if (data['status'] == 'error'):
+            failed.append({ 'url': data['url'], 'result': data['result'] })
+        else:
+            passed.append({ 'url': data['url'], 'result': data['result'] })
+
+    summary = {
+        'batchId': batchId,
+        'failed': {
+            'count': len(failed),
+            'results': failed
+        },
+        'passed': {
+            'count': len(passed),
+            'results': passed,
+        }
+    }
+
+    fileName = 'output.json'
+    s3URL = "s3://{}/{}/{}".format(bucketName, batchId, fileName)
+    with open(s3URL, 'w', transport_params={'client': session.client('s3')}) as fout:
+        json.dump(summary, fout)
+    
+    return {
+        "statusCode": 200,
+        "body": 'Processed batch: {}'.format(batchId)
+    }
+```
+
+The function retrieves the batch id from the http headers and uses it to iterate over the S3 Bucket's Contents. All results for the specific batch are read and added to the summary. As a final step the summary file called `output.json` is uploaded to the output folder for the batch.
+
+> When processing a large batch this function can take a while to complete. Make sure that your timeouts are configured correctly for both your function and the OpenFaaS core components. See: [Expanding timeouts](https://docs.openfaas.com/tutorials/expanded-timeouts/)
+
+## Conclusion
+We showed how a map/reduce pattern can be implemented with OpenFaaS. A workflow was created to process a CSV file containing Wikipedia URLs. Our goal was to run an machine learning model for each URL. To do this as quickly and efficiently as possible the input was split into many sub tasks that we could run in parallel. This fan out part is supported in OpenFaaS through asynchronous functions. After all the sub tasks completed their results were combined into a single output. This fanning in required some state. We used a redis key to keep track of the batch progress.
```

Comment:
We set out to show how the MapReduce pattern could be implemented with OpenFaaS, first by fanning out requests, using the built-in asynchronous system and queue-worker, then by fanning back in again by using shared storage.